controller/node_controller.go

```diff
@@ -1021,3 +871,215 @@ func BackingImageDiskFileCleanup(node *longhorn.Node, bi *longhorn.BackingImage,
+				}
+			} else {
+				// Clean up orphan CRs when
+				// 1. on-disk data is missing
+				// 2. on-disk orphaned data are on evicted or down disks
+				// 3. disk UUID is changed
+				id := orphan.Spec.Parameters[longhorn.OrphanDiskFsid]
+				replicaDirectoryName := orphan.Spec.Parameters[longhorn.OrphanDataName]
+				_, onDiskDataExist := onDiskReplicaDirectoryNames[id][replicaDirectoryName]
+
+				if !onDiskDataExist || isDiskDownOrEvicted(node, orphan) || isDiskUUIDChanged(node, orphan) {
+					if err := nc.ds.DeleteOrphan(orphan.Name); err != nil {
+						return errors.Wrapf(err, "failed to delete orphan %v", orphan.Name)
+					}
+				}
+			}
+		} else {
+			// Clean up orphan CRs whose owner node is down.
+			if err := nc.ds.DeleteOrphan(orphan.Name); err != nil {
+				return errors.Wrapf(err, "failed to delete orphan %v", orphan.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+func isDiskDownOrEvicted(node *longhorn.Node, orphan *longhorn.Orphan) bool {
+	id := orphan.Spec.Parameters[longhorn.OrphanDiskFsid]
+	disk, ok := node.Spec.Disks[id]
+	if !ok || (ok && disk.EvictionRequested) {
+		return true
+	}
+	return false
+}
+
+func isDiskUUIDChanged(node *longhorn.Node, orphan *longhorn.Orphan) bool {
+	id := orphan.Spec.Parameters[longhorn.OrphanDiskFsid]
+	if status, ok := node.Status.DiskStatus[id]; ok {
+		if orphan.Spec.Parameters[longhorn.OrphanDiskUUID] != status.DiskUUID {
+			return true
+		}
+	}
+	return false
+}
+
+func (nc *NodeController) createOrphans(node *longhorn.Node, onDiskReplicaDirectoryNames map[string]map[string]string) error {
+	// Find out the orphaned directories by checking with replica CRs
+	for diskID := range onDiskReplicaDirectoryNames {
+		for replicaName := range node.Status.DiskStatus[diskID].ScheduledReplica {
+			replica, err := nc.ds.GetReplica(replicaName)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					continue
+				} else {
+					return errors.Wrapf(err, "failed to get replica %v", replicaName)
+				}
+			}
+			delete(onDiskReplicaDirectoryNames[diskID], replica.Spec.DataDirectoryName)
+		}
+	}
+
+	// Now, the replica directories in onDiskReplicaDirectoryNames are orphaned, so we can create orphan resources for them.
+	for diskID, names := range onDiskReplicaDirectoryNames {
+		for name := range names {
+			if err := nc.createOrphan(node, name, diskID); err != nil && !apierrors.IsAlreadyExists(err) {
+				return errors.Wrapf(err, "unable to create orphan for orphaned replica directory %v in disk %v on node %v",
+					name, node.Spec.Disks[diskID].Path, node.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (nc *NodeController) createOrphan(node *longhorn.Node, replicaDirectoryName, id string) error {
+	orphanName := types.GetOrphanChecksumName(replicaDirectoryName, id, node.Name)
+
+	_, err := nc.ds.GetOrphanRO(orphanName)
+	if err == nil || (err != nil && !apierrors.IsNotFound(err)) {
+		return err
+	}
+
+	orphan := &longhorn.Orphan{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
I am thinking if adding `OwnershipReference` with cascading deletion would help simplify the logic.

Reply:
Can you elaborate it more?

Reply:
For example, after adding [the owner reference for backing image manager pods](https://github.com/longhorn/longhorn-manager/blob/51ee963e85e20a0a7b724c96fe67c876756ffd52/controller/backing_image_manager_controller.go#L782), the pod will be cleaned up when [the related backing image manager CR is removed](https://github.com/longhorn/longhorn-manager/blob/51ee963e85e20a0a7b724c96fe67c876756ffd52/datastore/longhorn.go#L1432).

Not sure if a similar logic can be applied to Orphan CRD vs Node CRD. Maybe the Orphan CRD is not node-specific in the future.